### PR TITLE
Decompose batch norm forward (based on #50)

### PR DESCRIPTION
**Pull request by kiya00**
This is based on the nvfuser in-place copy operator #50 . To improve readability for review, the modification related to batch_norm is presented here.

### CODE REVIEW

**Comment by kiya00** [comment]
After decomposing the bn forward pass, we get similar performance as main branch for the forward pass:
<details> <summary>decomposed bn fwd trace:</summary>

```
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast()
def computation(a, m, v, w, b):
  # a: "cuda:0 bf16[16, 128, 768]"
  # m: "cuda:0 bf16[128]"
  # v: "cuda:0 bf16[128]"
  # w: "cuda:0 bf16[128]"
  # b: "cuda:0 bf16[128]"
  [t45] = nvFusion0(a, b, m, v, w)
    # t1 = prims.convert_element_type(a, dtypes.float32)  # t1: "cuda:0 f32[16, 128, 768]"
    # (t4, t5) = prims.var_mean(t1, (0, 2), correction=0)
    # t6 = prims.broadcast_in_dim(t4, [1, 128, 1], [1])  # t6: "cuda:0 f32[1, 128, 1]"
    # t7 = prims.broadcast_in_dim(t5, [1, 128, 1], [1])  # t7: "cuda:0 f32[1, 128, 1]"
    # t8 = prims.add(t6, 1e-05)  # t8: "cuda:0 f32[1, 128, 1]"
    # t9 = prims.rsqrt(t8)  # t9: "cuda:0 f32[1, 128, 1]"
    # t10 = prims.broadcast_in_dim(t7, (16, 128, 768), (0, 1, 2))  # t10: "cuda:0 f32[16, 128, 768]"
    # t12 = prims.sub(t1, t10)  # t12: "cuda:0 f32[16, 128, 768]"
    # t13 = prims.broadcast_in_dim(t9, (16, 128, 768), (0, 1, 2))  # t13: "cuda:0 f32[16, 128, 768]"
    # t14 = prims.mul(t12, t13)  # t14: "cuda:0 f32[16, 128, 768]"
    # t15 = prims.squeeze(t7, (0, 2))  # t15: "cuda:0 f32[128]"
    # t16 = prims.convert_element_type(m, dtypes.float32)  # t16: "cuda:0 f32[128]"
    # t17 = prims.mul(0.8, t16)  # t17: "cuda:0 f32[128]"
    # t19 = prims.mul(0.2, t15)  # t19: "cuda:0 f32[128]"
    # t21 = prims.add(t17, t19)  # t21: "cuda:0 f32[128]"
    # t22 = prims.convert_element_type(t21, dtypes.bfloat16)  # t22: "cuda:0 bf16[128]"
    # prims.copy_(t22, m)
    # t24 = prims.squeeze(t6, (0, 2))  # t24: "cuda:0 f32[128]"
    # t25 = prims.mul(t24, 1.0000813868316107)  # t25: "cuda:0 f32[128]"
    # t26 = prims.convert_element_type(v, dtypes.float32)  # t26: "cuda:0 f32[128]"
    # t27 = prims.mul(0.8, t26)  # t27: "cuda:0 f32[128]"
    # t29 = prims.mul(0.2, t25)  # t29: "cuda:0 f32[128]"
    # t31 = prims.add(t27, t29)  # t31: "cuda:0 f32[128]"
    # t32 = prims.convert_element_type(t31, dtypes.bfloat16)  # t32: "cuda:0 bf16[128]"
    # prims.copy_(t32, v)
    # t34 = prims.reshape(w, (1, 128, 1))  # t34: "cuda:0 bf16[1, 128, 1]"
    # t35 = prims.broadcast_in_dim(t34, (16, 128, 768), (0, 1, 2))  # t35: "cuda:0 bf16[16, 128, 768]"
    # t36 = prims.convert_element_type(t35, dtypes.float32)  # t36: "cuda:0 f32[16, 128, 768]"
    # t37 = prims.mul(t14, t36)  # t37: "cuda:0 f32[16, 128, 768]"
    # t38 = prims.reshape(b, (1, 128, 1))  # t38: "cuda:0 bf16[1, 128, 1]"
    # t39 = prims.broadcast_in_dim(t38, (16, 128, 768), (0, 1, 2))  # t39: "cuda:0 bf16[16, 128, 768]"
    # t40 = prims.convert_element_type(t39, dtypes.float32)  # t40: "cuda:0 f32[16, 128, 768]"
    # t41 = prims.add(t37, t40)  # t41: "cuda:0 f32[16, 128, 768]"
    # t44 = prims.sum(t41, (0, 1, 2))  # t44: "cuda:0 f32[]"
    # t45 = prims.convert_element_type(t44, dtypes.bfloat16)  # t45: "cuda:0 bf16[]"
  del a, b, m, v, w
  return t45
```

</details>
<details> <summary>decomposed bn forward performance vs composed bn forward performance</summary>

decomposed:
image
composed:
image

</details>

But the manually written backward (main br) trace and the automatically generated backward trace(this pr) are quite different. Based on my initial benchmark testing, the automatically generated backward trace has worse performance
<details> <summary> The automatically generated backward trace(this pr)</summary>

```
@torch.no_grad()
@no_autocast()
def backward_fn(saved_for_backward, cotangents):
  # saved_for_backward: "Collection"
  # cotangents: "Collection"
  C0, _, = saved_for_backward
  clear_collection(saved_for_backward)
  del saved_for_backward
  t45, = cotangents
  clear_collection(cotangents)
  del cotangents
  a, t11, t34, t4, t8, = C0
  clear_collection(C0)
  del C0
  t154 = Tensor.expand(t45, (16, 128, 768))  # t154: "cuda:0 bf16[16, 128, 768]"
    # t154 = ltorch.expand(t45, (16, 128, 768))  # t154: "cuda:0 bf16[16, 128, 768]"
      # t154 = prims.broadcast_in_dim(t45, (16, 128, 768), ())  # t154: "cuda:0 bf16[16, 128, 768]"
  del t45
  t201 = torch.unsqueeze(t4, 0)  # t201: "cuda:0 f32[1, 128]"
    # t201 = ltorch.unsqueeze(t4, 0)  # t201: "cuda:0 f32[1, 128]"
      # t201 = prims.broadcast_in_dim(t4, [1, 128], [1])  # t201: "cuda:0 f32[1, 128]"
  del t4
  t202 = torch.unsqueeze(t201, 2)  # t202: "cuda:0 f32[1, 128, 1]"
    # t202 = ltorch.unsqueeze(t201, 2)  # t202: "cuda:0 f32[1, 128, 1]"
      # t202 = prims.broadcast_in_dim(t201, [1, 128, 1], [0, 1])  # t202: "cuda:0 f32[1, 128, 1]"
  del t201
  t188 = Tensor.expand(t202, [1, 128, 1])  # t188: "cuda:0 f32[1, 128, 1]"
    # t188 = ltorch.expand(t202, [1, 128, 1])  # t188: "cuda:0 f32[1, 128, 1]"
      # t188 = prims.broadcast_in_dim(t202, (1, 128, 1), (0, 1, 2))  # t188: "cuda:0 f32[1, 128, 1]"
  del t202
  t189 = Tensor.expand(t188, (16, 128, 768))  # t189: "cuda:0 f32[16, 128, 768]"
    # t189 = ltorch.expand(t188, (16, 128, 768))  # t189: "cuda:0 f32[16, 128, 768]"
      # t189 = prims.broadcast_in_dim(t188, (16, 128, 768), (0, 1, 2))  # t189: "cuda:0 f32[16, 128, 768]"
  del t188
  [t159, t167, t199] = nvFusion0(a, t11, t154, t189, t34, t8)
    # t0 = prims.convert_element_type(a, dtypes.float32)  # t0: "cuda:0 f32[16, 128, 768]"
    # t12 = prims.broadcast_in_dim(t8, (16, 128, 768), (0, 1, 2))  # t12: "cuda:0 f32[16, 128, 768]"
    # t13 = prims.mul(t11, t12)  # t13: "cuda:0 f32[16, 128, 768]"
    # t35 = prims.convert_element_type(t34, dtypes.float32)  # t35: "cuda:0 f32[16, 128, 768]"
    # t155 = prims.convert_element_type(t154, dtypes.float32)  # t155: "cuda:0 f32[16, 128, 768]"
    # t158 = prims.sum(t155, (0, 2))  # t158: "cuda:0 f32[128]"
    # t159 = prims.convert_element_type(t158, dtypes.bfloat16)  # t159: "cuda:0 bf16[128]"
    # t162 = prims.mul(t35, t155)  # t162: "cuda:0 f32[16, 128, 768]"
    # t163 = prims.mul(t13, t155)  # t163: "cuda:0 f32[16, 128, 768]"
    # t166 = prims.sum(t163, (0, 2))  # t166: "cuda:0 f32[128]"
    # t167 = prims.convert_element_type(t166, dtypes.bfloat16)  # t167: "cuda:0 bf16[128]"
    # t170 = prims.mul(t12, t162)  # t170: "cuda:0 f32[16, 128, 768]"
    # t171 = prims.mul(t11, t162)  # t171: "cuda:0 f32[16, 128, 768]"
    # t172 = prims.sum(t171, (0, 2))  # t172: "cuda:0 f32[128]"
    # t173 = prims.broadcast_in_dim(t172, [1, 128, 1], [1])  # t173: "cuda:0 f32[1, 128, 1]"
    # t174 = prims.neg(t170)  # t174: "cuda:0 f32[16, 128, 768]"
    # t176 = prims.sum(t174, (0, 2))  # t176: "cuda:0 f32[128]"
    # t177 = prims.broadcast_in_dim(t176, [1, 128, 1], [1])  # t177: "cuda:0 f32[1, 128, 1]"
    # t178 = prims.mul(-0.5, t173)  # t178: "cuda:0 f32[1, 128, 1]"
    # t179 = prims.pow(t8, 3.0)  # t179: "cuda:0 f32[1, 128, 1]"
    # t180 = prims.mul(t178, t179)  # t180: "cuda:0 f32[1, 128, 1]"
    # t181 = prims.sum(t177, (0, 2))  # t181: "cuda:0 f32[128]"
    # t182 = prims.sum(t180, (0, 2))  # t182: "cuda:0 f32[128]"
    # t183 = prims.broadcast_in_dim(t181, [1, 128, 1], [1])  # t183: "cuda:0 f32[1, 128, 1]"
    # t184 = prims.broadcast_in_dim(t183, (16, 128, 768), (0, 1, 2))  # t184: "cuda:0 f32[16, 128, 768]"
    # t185 = prims.mul(8.138020833333333e-05, t184)  # t185: "cuda:0 f32[16, 128, 768]"
    # t186 = prims.broadcast_in_dim(t182, [1, 128, 1], [1])  # t186: "cuda:0 f32[1, 128, 1]"
    # t187 = prims.broadcast_in_dim(t186, (16, 128, 768), (0, 1, 2))  # t187: "cuda:0 f32[16, 128, 768]"
    # t190 = prims.mul(2.0, t187)  # t190: "cuda:0 f32[16, 128, 768]"
    # t191 = prims.sub(t0, t189)  # t191: "cuda:0 f32[16, 128, 768]"
    # t192 = prims.mul(t190, t191)  # t192: "cuda:0 f32[16, 128, 768]"
    # t193 = prims.div(t192, 12288.0)  # t193: "cuda:0 f32[16, 128, 768]"
    # t194 = prims.add(t185, t193)  # t194: "cuda:0 f32[16, 128, 768]"
    # t198 = prims.add(t170, t194)  # t198: "cuda:0 f32[16, 128, 768]"
    # t199 = prims.convert_element_type(t198, dtypes.bfloat16)  # t199: "cuda:0 bf16[16, 128, 768]"
  del a, t11, t154, t189, t34, t8
  t205 = torch.unsqueeze(t159, 0)  # t205: "cuda:0 bf16[1, 128]"
    # t205 = ltorch.unsqueeze(t159, 0)  # t205: "cuda:0 bf16[1, 128]"
      # t205 = prims.broadcast_in_dim(t159, [1, 128], [1])  # t205: "cuda:0 bf16[1, 128]"
  del t159
  t206 = torch.unsqueeze(t205, 2)  # t206: "cuda:0 bf16[1, 128, 1]"
    # t206 = ltorch.unsqueeze(t205, 2)  # t206: "cuda:0 bf16[1, 128, 1]"
      # t206 = prims.broadcast_in_dim(t205, [1, 128, 1], [0, 1])  # t206: "cuda:0 bf16[1, 128, 1]"
  del t205
  t160 = Tensor.expand(t206, [1, 128, 1])  # t160: "cuda:0 bf16[1, 128, 1]"
    # t160 = ltorch.expand(t206, [1, 128, 1])  # t160: "cuda:0 bf16[1, 128, 1]"
      # t160 = prims.broadcast_in_dim(t206, (1, 128, 1), (0, 1, 2))  # t160: "cuda:0 bf16[1, 128, 1]"
  del t206
  t161 = torch.reshape(t160, (128,))  # t161: "cuda:0 bf16[128]"
    # t161 = ltorch.reshape(t160, (128,))  # t161: "cuda:0 bf16[128]"
      # t161 = prims.reshape(t160, (128,))  # t161: "cuda:0 bf16[128]"
  del t160
  t209 = torch.unsqueeze(t167, 0)  # t209: "cuda:0 bf16[1, 128]"
    # t209 = ltorch.unsqueeze(t167, 0)  # t209: "cuda:0 bf16[1, 128]"
      # t209 = prims.broadcast_in_dim(t167, [1, 128], [1])  # t209: "cuda:0 bf16[1, 128]"
  del t167
  t210 = torch.unsqueeze(t209, 2)  # t210: "cuda:0 bf16[1, 128, 1]"
    # t210 = ltorch.unsqueeze(t209, 2)  # t210: "cuda:0 bf16[1, 128, 1]"
      # t210 = prims.broadcast_in_dim(t209, [1, 128, 1], [0, 1])  # t210: "cuda:0 bf16[1, 128, 1]"
  del t209
  t168 = Tensor.expand(t210, [1, 128, 1])  # t168: "cuda:0 bf16[1, 128, 1]"
    # t168 = ltorch.expand(t210, [1, 128, 1])  # t168: "cuda:0 bf16[1, 128, 1]"
      # t168 = prims.broadcast_in_dim(t210, (1, 128, 1), (0, 1, 2))  # t168: "cuda:0 bf16[1, 128, 1]"
  del t210
  t169 = torch.reshape(t168, (128,))  # t169: "cuda:0 bf16[128]"
    # t169 = ltorch.reshape(t168, (128,))  # t169: "cuda:0 bf16[128]"
      # t169 = prims.reshape(t168, (128,))  # t169: "cuda:0 bf16[128]"
  del t168
  return (t199, None, None, t169, t161)
```

</details>

<details> <summary>The manually written backward trace(main br)</summary>

```
from torch import Tensor
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast()
def backward_fn(saved_for_backward, cotangents):
  # saved_for_backward: "Collection"
  # cotangents: "Collection"
  C0, C1, = saved_for_backward
  clear_collection(saved_for_backward)
  del saved_for_backward
  t10, = cotangents
  clear_collection(cotangents)
  del cotangents
  a, t4, t5, w, = C0
  clear_collection(C0)
  del C0
  b10, b6, b8, b9, f7, = C1
  clear_collection(C1)
  del C1
  t38 = Tensor.expand(t10, (16, 128, 768))  # t38: "cuda:0 bf16[16, 128, 768]"
    # t38 = ltorch.expand(t10, (16, 128, 768))  # t38: "cuda:0 bf16[16, 128, 768]"
      # t38 = prims.broadcast_in_dim(t10, (16, 128, 768), ())  # t38: "cuda:0 bf16[16, 128, 768]"
  del t10
  [t68, t69, t70] = nvFusion0(a, b10, b6, b8, b9, f7, t38, t4, t5, w)
    # t0 = prims.convert_element_type(a, dtypes.float32)  # t0: "cuda:0 f32[16, 128, 768]"
    # t1 = prims.convert_element_type(w, dtypes.float32)  # t1: "cuda:0 f32[128]"
    # t39 = prims.convert_element_type(t38, dtypes.float32)  # t39: "cuda:0 f32[16, 128, 768]"
    # t42 = prims.convert_element_type(t4, dtypes.float32)  # t42: "cuda:0 f32[128]"
    # t43 = prims.convert_element_type(t5, dtypes.float32)  # t43: "cuda:0 f32[128]"
    # t44 = prims.reshape(t42, (1, 128, 1))  # t44: "cuda:0 f32[1, 128, 1]"
    # t45 = prims.sum(t39, (0, 2))  # t45: "cuda:0 f32[128]"
    # t46 = prims.broadcast_in_dim(t44, (16, 128, 768), (0, 1, 2))  # t46: "cuda:0 f32[16, 128, 768]"
    # t47 = prims.sub(t0, t46)  # t47: "cuda:0 f32[16, 128, 768]"
    # t48 = prims.mul(t39, t47)  # t48: "cuda:0 f32[16, 128, 768]"
    # t49 = prims.sum(t48, (0, 2))  # t49: "cuda:0 f32[128]"
    # t50 = prims.mul(t45, 8.138020833333333e-05)  # t50: "cuda:0 f32[128]"
    # t51 = prims.reshape(t50, (1, 128, 1))  # t51: "cuda:0 f32[1, 128, 1]"
    # t52 = prims.mul(t49, 8.138020833333333e-05)  # t52: "cuda:0 f32[128]"
    # t53 = prims.mul(t43, t43)  # t53: "cuda:0 f32[128]"
    # t54 = prims.mul(t52, t53)  # t54: "cuda:0 f32[128]"
    # t55 = prims.reshape(t54, (1, 128, 1))  # t55: "cuda:0 f32[1, 128, 1]"
    # t56 = prims.mul(t43, t1)  # t56: "cuda:0 f32[128]"
    # t57 = prims.reshape(t56, (1, 128, 1))  # t57: "cuda:0 f32[1, 128, 1]"
    # t60 = prims.broadcast_in_dim(t55, (16, 128, 768), (0, 1, 2))  # t60: "cuda:0 f32[16, 128, 768]"
    # t61 = prims.mul(t47, t60)  # t61: "cuda:0 f32[16, 128, 768]"
    # t62 = prims.sub(t39, t61)  # t62: "cuda:0 f32[16, 128, 768]"
    # t63 = prims.broadcast_in_dim(t51, (16, 128, 768), (0, 1, 2))  # t63: "cuda:0 f32[16, 128, 768]"
    # t64 = prims.sub(t62, t63)  # t64: "cuda:0 f32[16, 128, 768]"
    # t65 = prims.broadcast_in_dim(t57, (16, 128, 768), (0, 1, 2))  # t65: "cuda:0 f32[16, 128, 768]"
    # t66 = prims.mul(t64, t65)  # t66: "cuda:0 f32[16, 128, 768]"
    # t67 = prims.mul(t49, t43)  # t67: "cuda:0 f32[128]"
    # t68 = prims.convert_element_type(t45, dtypes.bfloat16)  # t68: "cuda:0 bf16[128]"
    # t69 = prims.convert_element_type(t67, dtypes.bfloat16)  # t69: "cuda:0 bf16[128]"
    # t70 = prims.convert_element_type(t66, dtypes.bfloat16)  # t70: "cuda:0 bf16[16, 128, 768]"
  del a, b10, b6, b8, b9, f7, t38, t4, t5, w
  return (t70, None, None, t69, t68)
```

</details>

The patch I used for bn benchmark (PS: The performance result is flaky. I hope I used the benchmark correctly):
<details> <summary>The patch I used for bn benchmark:</summary>

```
diff --git a/thunder/benchmarks/__init__.py b/thunder/benchmarks/__init__.py
index 0c710cf..9d52214 100644
--- a/thunder/benchmarks/__init__.py
+++ b/thunder/benchmarks/__init__.py
@@ -2810,6 +2810,46 @@ class GPTBlockBenchmark(Benchmark, metaclass=UserFacingBenchmarkMeta):
         return model
 
 
+class BatchNormBenchmark(Benchmark, metaclass=UserFacingBenchmarkMeta):
+    def __init__(
+        self,
+        input_shape,
+        # batchdims: Sequence[int] = (16,),
+        device: str = "cuda",
+        dtype: dtypes.dtype = thunder.float32,
+        requires_grad: bool = False,
+    ) -> None:
+        super().__init__()
+
+        # self.config = _extract_nanogpt_config(config)
+        # self.batchdims = batchdims
+        self.shape: Sequence[int] = input_shape
+        self.device: str = device
+        self.dtype: dtypes.dtype = dtype
+        self.tdtype: torch.dtype = ltorch.to_torch_dtype(dtype)
+        self.requires_grad: bool = requires_grad
+
+        self.devices: list[str] = [device]
+
+    def make_batch(self) -> tuple[list, dict]:
+        make_arg = partial(make_tensor, device=self.device, dtype=self.tdtype, requires_grad=self.requires_grad)
+        make = partial(make_tensor, device=self.device, dtype=self.tdtype, requires_grad=False)
+        normalized_shape = (self.shape[1],)
+        a = make_arg(self.shape)
+        weight = make_arg(normalized_shape)# if w else None
+        bias = make_arg(normalized_shape)# if b else None
+        # 'batch_norm' is not differentiable with respect to argument 'running_mean' and 'running_var'
+        running_mean = make(normalized_shape)# if mean_var else None
+        running_var = make(normalized_shape)# if mean_var else None
+        # yield SampleInput(a, running_mean, running_var, weight, bias, training, **kwargs)
+        return (a, running_mean, running_var, weight, bias, True,), {}
+
+    def fn(self) -> Callable:
+        def foo(a,m,v,w,b, tr):
+            return torch.nn.functional.batch_norm(a,m,v,w,b,training=tr, momentum=0.2).sum()
+
+        return foo
+
 # TODO Add descriptions to the executors when listed, and list them alphabetically
 # TODO Allow querying benchmark for details
 # TODO Allow specifying benchmark arguments
diff --git a/thunder/benchmarks/targets.py b/thunder/benchmarks/targets.py
index 7bbe142..c364090 100644
--- a/thunder/benchmarks/targets.py
+++ b/thunder/benchmarks/targets.py
@@ -37,6 +37,7 @@ from thunder.benchmarks import (
     thunder_cudnn_layer_norm_nvfuser_executor,
     thunder_sdpa_executor,
     thunder_sdpa_torch_compile_nvfuser_executor,
+    BatchNormBenchmark,
 )
 
 from thunder.tests.litgpt_model import Config as LitGPTConfig
@@ -433,6 +434,43 @@ def test_nanogpt_gelu_grad(benchmark, executor: Callable):
     benchmark.pedantic(fn, setup=setup, rounds=40, warmup_rounds=1)
 
 
+@pytest.mark.parametrize(
+    "executor,",
+    fwd_executors,
+    ids=fwd_executor_ids,
+)
+def test_bn_fwd(benchmark, executor: Callable):
+    bn_bench: Benchmark = BatchNormBenchmark(
+        (16, 128, 768), device="cuda:0", dtype=thunder.bfloat16, requires_grad=False
+    )
+
+    setup = make_setup(bn_bench)
+    fn = executor(bn_bench)
+    fn = wrap_for_benchmark(fn)
+
+    benchmark.pedantic(fn, setup=setup, rounds=40, warmup_rounds=1)
+    # print(thunder.last_traces(fn)[-1])
+
+
+@pytest.mark.parametrize(
+    "executor,",
+    grad_executors,
+    ids=grad_executors_ids,
+)
+def test_bn_grad(benchmark, executor: Callable):
+    bn_bench: Benchmark = BatchNormBenchmark(
+        (16, 128, 768), device="cuda:0", dtype=thunder.bfloat16, requires_grad=True
+    )
+
+    setup = make_setup(bn_bench)
+    fn = executor(bn_bench)
+    fn = wrap_for_benchmark(fn)
+    benchmark.pedantic(fn, setup=setup, rounds=40, warmup_rounds=1)
+    # print(thunder.last_traces(fn)[-1])
+    # print(thunder.last_backward_traces(fn)[-1])
+
+
+
 # TODO Improve cross entropy's fwd+bwd perf when using the PyTorch executor
 #   See "torch.cross_entropy implementation has incorrect dtype metadata + bwd
 #        is very slow"
```

</details>

**Comment by IvanYashchuk** [comment]
"+99 −225" stats are pretty.

**Comment by IvanYashchuk** [comment]
> Based on my initial benchmark testing, the automatically generated backward trace has worse performance

Could you try disabling "bookend optimization" `nv_enable_bookend=False`?
https://github.com/Lightning-AI/lightning-thunder/blob/56c324c3dd11346c2b4c51946840c1f70d4cd45d/thunder/executors/nvfuserex_impl.py#L796C63-L796C80

**Review by IvanYashchuk — APPROVED** [review]
The main parts of the pull request look great!

For benchmarking, we should find out what is the most interesting shape to try comparing the performance between the old version and this one. Maybe increase the number of rounds and warmups to decrease the standard deviation.
Do we save the same number of tensors for backward now? Does bookend optimization have an impact on performance here? I think it should as there are many shape operations outside the nvFuser region in the automatically generated backward.

File: thunder/benchmarks/targets.py
Comment (kiya00) [reply thunder/benchmarks/targets.py:444]: Nothing particular, it comes from the shapes mentioned [here](https://github.com/Lightning-AI/lightning-thunder/blob/56c324c3dd11346c2b4c51946840c1f70d4cd45d/thunder/tests/opinfos.py#L6394) , I thought it comes from some networks

File: thunder/benchmarks/targets.py
Comment (kiya00) [reply thunder/benchmarks/targets.py:451]: I increase it to 200 rounds with 20 warmup_rounds, it's better but still not stable, and the same happens for torch.compile benchmark

**Comment by kiya00** [comment]
Hi @t-vi @mruberry , I think it's ready to merge

**Comment by jjsjann123** [comment]
> > > Based on my initial benchmark testing, the automatically generated backward trace has worse performance
> > 
> > 
> > Could you try disabling "bookend optimization" `nv_enable_bookend=False`? https://github.com/Lightning-AI/lightning-thunder/blob/56c324c3dd11346c2b4c51946840c1f70d4cd45d/thunder/executors/nvfuserex_impl.py#L796C63-L796C80
> 
> Currently the benchmark fails with `nv_enable_bookend=False`, and the error is the same as this [issue](https://github.com/NVIDIA/Fuser/issues/1964)
> 
> > Do we save the same number of tensors for backward now? Does bookend optimization have an impact on performance here? I think it should as there are many shape operations outside the nvFuser region in the automatically generated backward.
> 
> The automatically generated bwd is quite different from the manually written one, for a quick look that the manually written one saves 4 tensors for backward and now we save 5 tensors. It needs further analysis, should we use this decomposed forward and merge it for now? @jjsjann123 @IvanYashchuk

Thanks for trying out with bookend related issues! cc'ing @wujingyue 

Can you remind me of what's the motivation of decomposing batch norm? IIRC, better opportunity for rematerialization?! But looks like the benchmark is saying otherwise.
cc'ing @tfogal

**Comment by jjsjann123** [comment]
> The automatically generated bwd is quite different from the manually written one, for a quick look that the manually written one saves 4 tensors for backward and now we save 5 tensors. It needs further analysis, should we use this decomposed forward and merge it for now? 

I would prefer that we have some rough ideal on how to patch the regression to justify this PR before merging it. Otherwise it felt like we are back pedaling on our performance optimization endeavor.

I could be missing some context here, feel free to point out and push back :handshake:

**Comment by IvanYashchuk** [comment]
> Can you remind me of what's the motivation of decomposing batch norm? IIRC, better opportunity for rematerialization?! But looks like the benchmark is saying otherwise.

Thunder will have all operations decomposed to be recomposed by fusing executors so that it's more extensible and generalizable with great performance. There's no reason for batch norm decomposition to exist in nvFuser and not in Thunder. The benchmark doesn't say that this approach follows the wrong path, it says that there's a bug that needs to be fixed. All this PR is doing is lifting the batch norm definition from nvFuser to Thunder, nothing more nothing less. The generated CUDA kernels shouldn't be any different.

**Comment by jjsjann123** [comment]
> > Can you remind me of what's the motivation of decomposing batch norm? IIRC, better opportunity for rematerialization?! But looks like the benchmark is saying otherwise.
> 
> Thunder will have all operations decomposed to be recomposed by fusing executors so that it's more extensible and generalizable with great performance. There's no reason for batch norm decomposition to exist in nvFuser and not in Thunder. The benchmark doesn't say that this approach follows the wrong path, it says that there's a bug that needs to be fixed. All this PR is doing is lifting the batch norm definition from nvFuser to Thunder, nothing more nothing less. The generated CUDA kernels shouldn't be any different.

Sounds like it's just nvfuser not doing the right partitioning... :cry: 
Looks like #182 has it patched so we are all good here. I'm not too paranoid about the order of merge here.

:shipit:

**Comment by IvanYashchuk** [comment]
@t-vi, a code owner approval is needed for merging. Thanks!

**Pull request merged**
